### PR TITLE
Skip synchronization of outer positions on writes

### PR DESCRIPTION
**Pull request by pnowojski**
Addresses this issue: https://github.com/prestodb/presto/issues/6923

Not an ideal solution, better one would be to completely omit synchronization by using one copy of `visitedPosition` per thread and later merge them together, but from benchmarking this already gives significant improvement (this one is still ~30% slower compared to version without synchronization before @dain changes) . 

CC @KBP-TDC

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java [hunk 1]
```diff
@@ -197,38 +203,24 @@ public synchronized boolean appendToNext(PageBuilder pageBuilder, int outputChan
...
-            for (int source = 0; source < lookupSourceSuppliers.size(); source++) {
-                try (LookupSource lookupSource = lookupSourceSuppliers.get(source).get()) {
-                    visitedPositions[source] = new boolean[lookupSource.getJoinPositionCount()];
-                }
-            }
+            visitedPositions = new boolean[positionCount];
         }
 
-        public synchronized void positionVisited(int partition, int position)
+        public synchronized void positionVisited(int position)
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java]: I do not like exposing visitedPositions arrays from position trackers.
I know this is stop-gap but it seems it can be done simpler by changing
`positionVisited` to:
```java
public void positionVisited(int position) {
    verify(!finished);
    synchronised(visitedPositions[partition]) {
             visitedPositions[partition][position] = true;
    }
}
```

And `getOuterPositionIterator` to:
 
```java
public OuterPositionIterator getOuterPositionIterator() {
    finished = true;
    LookupSource[] lookupSources = lookupSourceSuppliers.stream()           
                     .map(Supplier::get)		
                     .toArray(LookupSource[]::new);
     for (int i = 0; i < visitedPartitions.size; ++i) {
        synchronized(visitedPartitions[i]) {} // just a memory barrier
     }
 		
      return new PartitionedLookupOuterPositionIterator(lookupSources, visitedPositions);
}
```

`finished` flag would probably need to be made volatile.

**Comment by pnowojski** [comment]
@dain @losipiuk here is updated version, with only one synchronization per thread in `commit` method. Our benchmarking cluster is down for a maintenance, but I expect to see this to completely fix https://github.com/prestodb/presto/issues/6923

File: presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java [hunk 2]
```diff
@@ -174,11 +174,11 @@ public void close()
         closed = true;
         probe = null;
         pageBuilder.reset();
-        onClose.run();
         // closing lookup source is only here for index join
         if (lookupSource != null) {
             lookupSource.close();
         }
+        onClose.run();
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java]: why this change?

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java [hunk 3]
```diff
@@ -194,41 +206,42 @@ public synchronized boolean appendToNext(PageBuilder pageBuilder, int outputChan
...
             visitedPositions[partition][position] = true;
         }
 
-        public synchronized OuterPositionIterator getOuterPositionIterator()
+        public synchronized void commit()
+        {
+            committed = true;
+        }
+
+        public synchronized boolean[][] getVisitedPositions()
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java]: I do not see how this code enforces that
getVisitedPositions is called only after _every_ other thread went through synchronized section on `OuterPositionTracker`.
Lets discuss this offline.

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java [hunk 4]
```diff
@@ -194,41 +206,82 @@ public synchronized boolean appendToNext(PageBuilder pageBuilder, int outputChan
         }
     }
 
-    @ThreadSafe
+    /**
+     * Each thread has it's own copy of OuterPositionTracker instance and after using it,
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java]: It should rather say that each LookupSource has its own OPT. And that last thread which uses it must commit it.

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java [hunk 5]
```diff
@@ -40,14 +40,19 @@
 {
     public static Supplier<LookupSource> createPartitionedLookupSourceSupplier(List<Supplier<LookupSource>> partitions, List<Type> hashChannelTypes, boolean outer)
     {
-        Optional<OuterPositionTracker> outerPositionTracker = Optional.ofNullable(outer ? new OuterPositionTracker(partitions) : null);
+        Optional<OuterPositionTracker.OuterPositionTrackerFactory> outerPositionTracker = outer ?
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java]: name variable `outerPositionTrackerFactory`

Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java]: fixed

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java (hunk 4 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java]: fixed

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java [hunk 6]
```diff
@@ -194,41 +206,82 @@ public synchronized boolean appendToNext(PageBuilder pageBuilder, int outputChan
...
+     * visitedPositions, writes can be without synchronization.
+     *
+     * Synchronization in commit() is required because each writing thread must have
+     * memory barrier, so that it writes its state of visitedPositions to main memory
+     * before visitedPositions is read in getVisitedPositions().
+     */
     private static class OuterPositionTracker
     {
-        private final List<Supplier<LookupSource>> lookupSourceSuppliers;
+        public static class OuterPositionTrackerFactory
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java]: name this simply `Factory`

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java [hunk 7]
```diff
@@ -194,41 +206,82 @@ public synchronized boolean appendToNext(PageBuilder pageBuilder, int outputChan
...
+            {
+                visitedPositions = new boolean[positionCounts.size()][];
+                for (int partition = 0; partition < visitedPositions.length; partition++) {
+                    visitedPositions[partition] = new boolean[positionCounts.get(partition)];
+                }
+            }
 
-        @GuardedBy("this")
-        private final boolean[][] visitedPositions;
+            public synchronized OuterPositionTracker createOuterPositionTracker()
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java]: rename this to `create`

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java (hunk 6 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java]: fixed

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java (hunk 7 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java]: fixed

**Comment by pnowojski** [comment]
I have run some benchmarks and with this patch there is no noticeable difference between INNER JOIN and FULL OUTER JOIN:
```
presto:tpcds_1tb_orc> SELECT COUNT(*) FROM store_sales s FULL OUTER JOIN web_sales w ON s.ss_item_sk = w.ws_item_sk AND s.ss_sold_date_sk = w.ws_sold_date_sk;
   _col0    
------------
 9741265857 
(1 row)

Query 20161229_102047_00075_jhrnd, FINISHED, 8 nodes
Splits: 3,513 total, 3,513 done (100.00%)
1:39 [3.6B rows, 197GB] [36.3M rows/s, 1.98GB/s]

presto:tpcds_1tb_orc> SELECT COUNT(*) FROM store_sales s INNER JOIN web_sales w ON s.ss_item_sk = w.ws_item_sk AND s.ss_sold_date_sk = w.ws_sold_date_sk;
   _col0    
------------
 9375443834 
(1 row)

Query 20161229_102229_00076_jhrnd, FINISHED, 8 nodes
Splits: 3,505 total, 3,505 done (100.00%)
1:37 [3.6B rows, 197GB] [37.1M rows/s, 2.03GB/s]
```

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java [hunk 8]
```diff
@@ -194,41 +206,83 @@ public synchronized boolean appendToNext(PageBuilder pageBuilder, int outputChan
...
         }
 
-        public synchronized void positionVisited(int partition, int position)
+        /**
+         * No synchronization here, because it would be very expensive. Check comment above.
+         */
+        public void positionVisited(int partition, int position)
         {
-            verify(!finished);
+            if (!written) {
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java]: Why do we need to keep track of whether it was written to or not? Wouldn't it be sufficient to increment the reference count in the constructor and decrement it unconditionally when commit() is called?

Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java]: AFAIK that was initial approach but there are cases when tracker is created but `commit()` never called, when tracker is not actually used for updating `visitedPositions` array

Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java]: Which cases? That would imply that LookupSource.close() is not being called in some cases, which seems like a broken interaction. If a LookupSource is created, it should always be closed.

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java [hunk 9]
```diff
@@ -194,41 +206,83 @@ public synchronized boolean appendToNext(PageBuilder pageBuilder, int outputChan
...
+     * Each LookupSource has it's own copy of OuterPositionTracker instance.
+     * Each of those OuterPositionTracker must be committed after last write
+     * and before first read.
+     *
+     * All instances share visitedPositions array, but it is safe because each thread
+     * starts with visitedPositions filled with false values and marks only some positions
+     * to true. Since we don't care what will be the order of those writes to
+     * visitedPositions, writes can be without synchronization.
+     *
+     * Synchronization in commit() is required because each writing thread must have
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java]: Does JMM say anything about Atomics explicitly?
Or just that write to volatile variable enforces happens-before semantic related to any subsequent read of same variable?

It is probably a safe assumption that AtomicLong uses volatile internally. Yet isn't that more explicit to just use monitor here which also enforces happens-before semantics between actions in threads entering it?

The correctness is then assured by the facts that:
1. All mutations to `vistitedPositions` are happening in each thread before this thread calls commit()
1. Each thread calls `commit()` after it performed all the mutations to `visitedPositions`
1. `getVisitedPositions` is called after all the threads which performed mutations to `visitedPositions` call `commit()`
1. commit() and `getVisitedPositions()` share monitor

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java (hunk 8 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java]: I will let @pnowojski reply. He digged deeper into that.

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java (hunk 9 above)
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java]: > Does JMM say anything about Atomics explicitly?

Yes, is explicit about it:

> The memory effects for accesses and updates of atomics generally follow the rules for volatiles, as stated in section 17.4 of The Java™ Language Specification.
> * **get** has the memory effects of reading a volatile variable.
> * **set** has the memory effects of writing (assigning) a volatile variable.
> * **lazySet** has the memory effects of writing (assigning) a volatile variable except that it permits reorderings with subsequent (but not previous) memory actions that do not themselves impose reordering  constraints with ordinary non-volatile writes. Among other usage contexts, lazySet may apply when nulling out, for the sake of garbage collection, a reference that is never accessed again.
> * **weakCompareAndSet** atomically reads and conditionally writes a variable but does not create any happens-before orderings, so provides no guarantees with respect to previous or subsequent reads and writes of any variables other than the target of the weakCompareAndSet.
> * **compareAndSet** and all other read-and-update operations such as getAndIncrement have the memory effects of both reading and writing volatile variables.

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java [hunk 10]
```diff
@@ -194,41 +206,83 @@ public synchronized boolean appendToNext(PageBuilder pageBuilder, int outputChan
...
+     * memory barrier, so that it writes its state of visitedPositions to main memory
+     * before visitedPositions is read in getVisitedPositions().
+     */
     private static class OuterPositionTracker
     {
-        private final List<Supplier<LookupSource>> lookupSourceSuppliers;
+        public static class Factory
+        {
+            private final boolean[][] visitedPositions;
+            private AtomicLong referenceCount = new AtomicLong();
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java]: fixed

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java (hunk 9 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java]: @martint dropping those synchronizations would indeed work just as fine, as long as we keep this `AtomicLong referenceCount` variable. However it is an implementation detail whether `referenceCount` exists or not. In this case. It is there only for the sanity check done later:

```verify(referenceCount.get() == 0);```

We could delete `referenceCount` and code would work just as fine. By dropping synchronization in `commit` and `getVisitedPositions` I would have to add comment to `referenceCount` variable: "DO NOT REMOVE IT, because it's synchronization is required to guarantee memory visibility between `commit` and `getVisitedPositions`". I think that would be less explicit and more hacky compared to `synchronized` methods. And it wouldn't give any performance improvement.

Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java]: >  I would have to add comment to referenceCount variable: "DO NOT REMOVE IT, because it's synchronization is required to guarantee memory visibility between commit and getVisitedPositions"

You'd have to do that for `synchronized` on the method, regardless. Otherwise, an empty synchronized method would look odd and someone might end up removing it.

I think the reference count is a smart way of encoding the required semantics:
* If `getVisitedPositions` saw all the decrements, then it must see the updates to the positions.
* If `getVisitedPositions` didn't see all the decrements, there's a sequencing bug and the method will fail loudly.

Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java]: ok

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java (hunk 8 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java]: There might be some unrelated bugs with missing close calls, but the main reason is that to get `OuterPositionIterator` we currently need to create `LookupSource` and thus `OuterPositionTracker`. So `referenceCount` incremented in constructor would never be `0`.

I don't like all the relations between all of those classes and their interactions, but I didn't see an easy clear way how to straighten all the problems and I didn't want to dig in into refactor attempt as a part of this regression fix.

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java [hunk 11]
```diff
@@ -194,41 +206,83 @@ public synchronized boolean appendToNext(PageBuilder pageBuilder, int outputChan
...
+            {
+                visitedPositions = new boolean[positionCounts.size()][];
+                for (int partition = 0; partition < visitedPositions.length; partition++) {
+                    visitedPositions[partition] = new boolean[positionCounts.get(partition)];
+                }
+            }
 
-        @GuardedBy("this")
-        private final boolean[][] visitedPositions;
+            public synchronized OuterPositionTracker create()
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java]: fixed

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java [hunk 12]
```diff
@@ -194,41 +206,83 @@ public synchronized boolean appendToNext(PageBuilder pageBuilder, int outputChan
...
+            }
             visitedPositions[partition][position] = true;
         }
 
-        public synchronized OuterPositionIterator getOuterPositionIterator()
+        /**
+         * Synchronization only for ensuring that there is a memory barrier before after writting.
+         * Check comment above.
+         */
+        public synchronized void commit()
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java]: dropped

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java [hunk 13]
```diff
@@ -194,41 +206,83 @@ public synchronized boolean appendToNext(PageBuilder pageBuilder, int outputChan
...
+        {
+            if (written) {
+                referenceCount.decrementAndGet();
+            }
+        }
+
+        public synchronized boolean[][] getVisitedPositions()
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java]: dropped

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java (hunk 8 above)
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java]: > the main reason is that to get OuterPositionIterator we currently need to create LookupSource and thus OuterPositionTracker. So referenceCount incremented in constructor would never be 0.

But that should be fine, since the OuterPositionIterator is created by PartitionedLookupSource, which already has the LookupSources.

I'll get this in as is, but we should investigate why we're missing some calls to close().

**Comment by martint** [comment]
Merged, thanks!

**Comment by pnowojski** [comment]
Thanks
